product_supplierinfo_for_customer/models/product_product.py

```diff
@@ -130,10 +130,25 @@ def _select_customerinfo(
             params = dict()
         params.update({"partner_id": partner.id})
         domain = self._prepare_domain_customerinfo(params)
-        res = (
+        # Search for customerinfo records based on the domain
+        any_match = (
             self.env["product.customerinfo"]
             .search(domain)
             .sorted(lambda s: (s.sequence, s.min_qty, s.price, s.id))
         )
-        res_1 = res.sorted("product_tmpl_id")[:1]
-        return res_1
+        # First, try to find a matching variant
+        variant_match = any_match.filtered(
+            lambda r: r.product_id and r.product_id == self
+        )
+        if variant_match:
+            return variant_match[:1]  # Return the first matching variant if any
+        # If no matching variant, then look for matches based on template
+        template_match = any_match.filtered(
+            lambda r: r.product_tmpl_id
+            and r.product_tmpl_id == self.product_tmpl_id
+            and not r.product_id
+        )
+        if template_match:
+            return template_match[:1]
+        any_match_1 = any_match.sorted("product_tmpl_id")[:1]
+        return any_match_1
```

Comment:
IMO All this could be done in one loop....
```suggestion
        first_template_match = None
        # return the first customer_info matching the variant
        # otherwise the first one matching the template
        for customer_info in any_match:
            if customer_info.product_id == self:
                return customer_info
            if not customer_info.product_id and \
                    not first_template_match and \
                    customer_info.product_tmpl_id == self.product_tmpl_id:
                first_template_match = customer_info
        return first_template_match
```

Reply:
Much better, thanks